### PR TITLE
Fix broken DNS native_server

### PR DESCRIPTION
**Pull request by smashery**
This fixes two bugs in the `auxiliary/server/dns/native_server` module:

- It would immediately crash upon receiving any request
- Once that was fixed, if a request contained two queries, but only one was served from the cache and another from upstream, they were combined into the question field of the reply, not the answer.

Demo of the first issue (send any DNS request to the server):

```
msf6 > use auxiliary/server/dns/native_server                                                                      
msf6 auxiliary(server/dns/native_server) > set srvhost 0.0.0.0                                                     
srvhost => 0.0.0.0                                                                                                                                                                                                                   
msf6 auxiliary(server/dns/native_server) > set static_entries "1.2.3.4 example.com"                                
static_entries => 1.2.3.4 example.com              
msf6 auxiliary(server/dns/native_server) > run                                                                    
                                                                                                                  
[-] Auxiliary failed: NoMethodError undefined method `data' for #<Dnsruby::Message:0x000055d763109448>
[-] Call stack:                                                                                                   
[-]   /home/smash/git/metasploit-framework/lib/rex/proto/dns/resolver.rb:136:in `send'
[-]   /home/smash/git/metasploit-framework/modules/auxiliary/server/dns/native_server.rb:80:in `on_dispatch_request'
[-]   /home/smash/git/metasploit-framework/lib/msf/core/exploit/remote/dns/server.rb:118:in `block in start_service'
[-]   /home/smash/git/metasploit-framework/lib/rex/proto/dns/server.rb:273:in `dispatch_request'
[-]   /home/smash/git/metasploit-framework/lib/rex/proto/dns/server.rb:349:in `monitor_listener'
[-]   /home/smash/git/metasploit-framework/lib/rex/proto/dns/server.rb:230:in `block in start'
[-]   /home/smash/git/metasploit-framework/lib/rex/thread_factory.rb:22:in `block in spawn'
[-]   /home/smash/git/metasploit-framework/lib/msf/core/thread_manager.rb:105:in `block in spawn'
[-]   /var/lib/gems/2.7.0/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

To my reading, this code will always work with a DnsRuby object (which expects the `:encode` method), not a `Net::DNS` object (which expects the `:data` method). If a Net::DNS object is passed in, for instance, it will be converted to a DnsRuby object first.

The second issue I just saw directly when reviewing the code; it seems that a copy paste error occurred in this change:

https://github.com/smashery/metasploit-framework/commit/c65c03722c89428e1233ad2304731835d27bb4a5#diff-eccd9746aa08a91a52512ad5b324483719011ba6627693e7bee8323d8eadcc2eL84-R86

It's probably a super rare situation... to be honest, I can't even figure out how to test it without pulling out scapy or something.

### CODE REVIEW

File: lib/rex/proto/dns/resolver.rb [hunk 1]
```diff
@@ -132,8 +132,8 @@ def send(argument, type = Dnsruby::Types::A, cls = Dnsruby::Classes::IN)
       end
 
       # Store packet_data for performance improvements,
-      # so methods don't keep on calling Packet#data
-      packet_data = packet.data
+      # so methods don't keep on calling Packet#encode
+      packet_data = packet.encode
```
Comment (smcintyre-r7) [inline lib/rex/proto/dns/resolver.rb:136]: This unfortunately introduces a bug in the `auxiliary/gather/enum_dns` module which I'm assuming uses a different object type for `packet`.

```
msf6 auxiliary(gather/enum_dns) > run

[-] Auxiliary failed: NoMethodError undefined method `encode' for ;; HEADER SECTION
;; id = 61527
;; qr = 0	opCode: QUERY	aa = 0	tc = 0	rd = 1
;; ra = 0	ad = 0	cd = 0	rcode = NoError
;; qdCount = 1	anCount = 0	nsCount = 0	arCount = 0

;; QUESTION SECTION (1 record):
;; zpzmfhGQfHxeVuel.home.lan    IN      A       
:Net::DNS::Packet
Did you mean?  encode_tlv
[-] Call stack:
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/proto/dns/resolver.rb:136:in `send'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/net/dns/resolver.rb:894:in `query'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/dns/client.rb:61:in `query'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/dns/enumeration.rb:118:in `dns_get_a'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/dns/enumeration.rb:366:in `dns_wildcard_enabled?'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/gather/enum_dns.rb:71:in `run'
[*] Auxiliary module execution completed
```

**Review by smcintyre-r7 — COMMENTED** [review]
I went to test this and I was not able to reproduce the original issue. With STATIC_ENTRIES set, as in your description, the module seems to be working for me on the master branch. I'm wondering if for some reason my instance using using Net::Dns while you're using Dnsruby or vice versa.

File: lib/rex/proto/dns/resolver.rb (hunk 1 above)
Comment (smcintyre-r7) [reply lib/rex/proto/dns/resolver.rb:136]: Had a chance to look at this a bit more thorougly. The lines just above that were hidden by GitHub make it clear that this code path is intended to handle messages from both `Net::DNS` and `Dnsruby`. Looks like this change should be merged into that block making it:

```ruby
      case argument
      when Dnsruby::Message
        packet = argument
        packet_data = packet.encode
      when Net::DNS::Packet, Resolv::DNS::Message
        packet = Rex::Proto::DNS::Packet.encode_drb(argument)
        packet_data = packet.data
      else
        packet = make_query_packet(argument,type,cls)
        packet_data = packet.data
      end

      # Store packet_data for performance improvements,
      # so methods don't keep on calling Packet#data
      packet_size = packet_data.size
```

[make_query_packet](https://github.com/bluemonk/net-dns/blob/21d32c0b2e54ccdd3f9f27c53fe518316bf719f9/lib/net/dns/resolver.rb) is provided by Net::DNS and should return a Net::DNS object, meaning `#data` is the appropriate method to use in this case.

Comment (smcintyre-r7) [reply lib/rex/proto/dns/resolver.rb:136]: This change also fixed the issue I was running into with the `enum_dns` module. I did have to work around another, unrelated issue by setting `DnsNote` to false though.

Comment (smcintyre-r7) [reply lib/rex/proto/dns/resolver.rb:136]: Fixed the unrelated `enum_dns` issue in #16347 that I noticed while testing these changes to ensure that that was no regression.

Comment (smashery) [reply lib/rex/proto/dns/resolver.rb:136]: In the second `when` statement, the `encode_drb` returns a `Dnsruby::Message` object, so I think `encode` will be correct there. I think the only issue there is in the third one, which, as you say, returns a `Net::DNS::Packet` object.

**Review by smcintyre-r7 — APPROVED** [review]
I was able to confirm that the native server module is now working!

<details>
<summary>Testing Output</summary>

```
msf6 auxiliary(gather/enum_dns) > use auxiliary/server/dns/native_server 
msf6 auxiliary(server/dns/native_server) > show options 

Module options (auxiliary/server/dns/native_server):

   Name              Current Setting      Required  Description
   ----              ---------------      --------  -----------
   DISABLE_NS_CACHE  false                no        Disable DNS response caching
   DISABLE_RESOLVER  false                no        Disable DNS request forwarding
   DOMAIN                                 no        The target domain name
   NS                                     no        Specify the nameservers to use for queries, space separated
   Proxies                                no        A proxy chain of format type:host:port[,type:host:port][...]
   RPORT             53                   yes       The target port (TCP)
   SEARCHLIST                             no        DNS domain search list, comma separated
   SRVHOST           192.168.159.128      yes       The local host or network interface to listen on. This must be
                                                    an address on the local machine or 0.0.0.0 to listen on all add
                                                    resses.
   SRVPORT           53                   yes       The local port to listen on.
   STATIC_ENTRIES    1.2.3.4 example.com  no        DNS domain search list (hosts file or space/semicolon separate
                                                    entries)
   THREADS           1                    yes       Number of threads to use in threaded queries


Auxiliary action:

   Name     Description
   ----     -----------
   Service  Serve DNS entries


msf6 auxiliary(server/dns/native_server) > run
[*] Auxiliary module running as background job 0.
msf6 auxiliary(server/dns/native_server) > dig example.com @192.168.159.128
[*] exec: dig example.com @192.168.159.128


; <<>> DiG 9.16.24-RH <<>> example.com @192.168.159.128
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 37299
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;example.com.			IN	A

;; ANSWER SECTION:
example.com.		0	IN	A	1.2.3.4

;; Query time: 2 msec
;; SERVER: 192.168.159.128#53(192.168.159.128)
;; WHEN: Tue Mar 15 09:04:34 EDT 2022
;; MSG SIZE  rcvd: 45

msf6 auxiliary(server/dns/native_server) >
```
</details>

**Comment by smcintyre-r7** [comment]
# Release Notes
This fixes an issue in the DNS native server module where by the server would crash upon receiving a query.

**Pull request merged**